### PR TITLE
556：component pattern

### PR DESCRIPTION
**Pull request by Hrighter**
Pull request description
- for issue: [#556](https://github.com/iluwatar/java-design-patterns/issues/556)
- create a component pattern demo

### CODE REVIEW

File: component/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
+layout: pattern  
+title: Component  
+folder: component  
+permalink: /patterns/component/  
+categories: Behavioral  
+tags:  
+ - Java
```
Comment (iluwatar) [inline component/README.md:8]: I would suggest `Game programming` and `Decoupling` tags. See https://github.com/iluwatar/java-design-patterns/wiki/07.-Categories-and-Tags

File: component/pom.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        <version>1.23.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>component</artifactId>
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
```
Comment (iluwatar) [inline component/pom.xml:17]: You'll get this from the parent `pom.xml`. No need to define the version here.

File: component/src/main/java/com/iluwater/component/App.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * 1.the coupling of data organization is very strong.
+ * 2.Interface logic is difficult to reuse and hot plug.
+ *
+ * The component pattern solves the defects of object orientation and process orientation and
+ * is widely used in game clients
+ *
+ * A component is a part of one object. We can consider that a object contains multiple
+ * components, in another way, multiple components can construct a object.
+ * Here is a demo using component pattern to solve a game-like problem.
+ * A person named Bjorn who has three component: get input;judge physics state;show graphics state
```
Comment (iluwatar) [inline component/src/main/java/com/iluwater/component/App.java:39]: component -> components

File: component/src/main/java/com/iluwater/component/App.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * These three component with a common game object can construct our protagonist:Bjorn, also can
```
Comment (iluwatar) [inline component/src/main/java/com/iluwater/component/App.java:40]: component -> components

File: component/src/main/java/com/iluwater/component/App.java (hunk 3 above)
Comment (iluwatar) [reply component/src/main/java/com/iluwater/component/App.java:39]: get input;judge physics state;show graphics state -> input, physics, graphics

File: component/src/main/java/com/iluwater/component/App.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * construct other objects like dog or cat if you want to write a real game.
+ */
+
+public class App {
+    public static void main(String[] args) {
+        ArrayList<Component>arrayList = new ArrayList<>();
```
Comment (iluwatar) [inline component/src/main/java/com/iluwater/component/App.java:46]: We can use Java 11 `var` here to declare variables. Check elsewhere in the code too.

File: component/src/main/java/com/iluwater/component/GameObject.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+package com.iluwater.component;
+
+import java.util.ArrayList;
+
+public class GameObject {
+    public int velocity;
+    public int positionOFx;
+    public int positionOFy;
```
Comment (iluwatar) [inline component/src/main/java/com/iluwater/component/GameObject.java:8]: Should be private

File: component/src/main/java/com/iluwater/component/GameObject.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.ArrayList;
+
+public class GameObject {
+    public int velocity;
+    public int positionOFx;
+    public int positionOFy;
+    ArrayList<Component> componentArrayList;
+    public GameObject(ArrayList<Component> componentArrayList){
+        this.componentArrayList = componentArrayList;
```
Comment (iluwatar) [inline component/src/main/java/com/iluwater/component/GameObject.java:11]: Instead of assigning the array I would copy the contents

**Comment by Hrighter** [comment]
I will fix these problems soon. Thank you for your review.

**Comment by Hrighter** [comment]
I have fixed these problems, but one thing is that for some reason I can't use the checkstyle file in the project, so I use the built-in sun checkstyle file to test.

**Comment by Hrighter** [comment]
Sorry for my wrong operation when I try to create a new branch to create a new pull request for another issue,  I will close this one and then create two new pull requests.

**Pull request closed without merging**